// react-native-firebase/dist/modules/database/disconnect.js
/**
 * 
 * Disconnect representation wrapper
 */
import { typeOf } from './../../utils';


/**
 * @url https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect
 * @class Disconnect
 */
export default class Disconnect {

  /**
   *
   * @param ref
   */
  constructor(ref) {
    this.ref = ref;
    this.path = ref.path;
    this._database = ref._database;
  }

  /**
   * @url https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#set
   * @param value
   * @returns {*}
   */
  set(value) {
    return this._database._native.onDisconnectSet(this.path, { type: typeOf(value), value });
  }

  /**
   * @url https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#update
   * @param values
   * @returns {*}
   */
  update(values) {
    return this._database._native.onDisconnectUpdate(this.path, values);
  }

  /**
   * @url https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#remove
   * @returns {*}
   */
  remove() {
    return this._database._native.onDisconnectRemove(this.path);
  }

  /**
   * @url https://firebase.google.com/docs/reference/js/firebase.database.OnDisconnect#cancel
   * @returns {*}
   */
  cancel() {
    return this._database._native.onDisconnectCancel(this.path);
  }
}
